File: patch_shared_calc.js
const fs = require('fs');
const path = 'public/sejong/shared_calc.js';
let content = fs.readFileSync(path, 'utf8');

// 1. Move getCycle up
content = content.replace(/const getCycle = \(val\) => \{[\s\S]*?\};\s*let currentCycle = getCycle\(currentMC\.carryFromPrev\);\s*if \(isNaN\(currentCycle\)\) currentCycle = 0;/, '');

const newCycleLogic = `
    const getCycle = (val) => {
        let vRaw = Math.round(val * 10);
        if (isDualCourse) {
            if (vRaw < 170) return 0;
            return Math.floor((vRaw - 170) / 160) + 1;
        } else {
            if (vRaw < 90) return 0;
            return Math.floor((vRaw - 90) / 80) + 1;
        }
    };

    let globalRunningTotal = 0;
    let globalCurrentCycle = 0;
    let allMilestones = [];
`;

content = content.replace(/let carryOverP = 0;/, newCycleLogic + '\n    let carryOverP = 0;');

// 2. Track milestones inside monthsToCalc.forEach > mLogs.forEach
const oldMLogsForEach = `
            if (isMakeupMarker) manualMakeup += attendanceIncrement;

            if (isRegularAttendance) {
                attendances += attendanceIncrement;
            }
        });`;

const newMLogsForEach = `
            if (isMakeupMarker) manualMakeup += attendanceIncrement;

            if (isRegularAttendance) {
                attendances += attendanceIncrement;
            }

            if (isRegularAttendance || isMakeupMarker) {
                const prevNet = globalRunningTotal;
                globalRunningTotal += attendanceIncrement;
                globalRunningTotal = Math.round(globalRunningTotal * 10) / 10;
                const prevCycle = getCycle(prevNet);
                const currCycle = getCycle(globalRunningTotal);
                
                const dateStr = l.date.split('T')[0];
                const isForced = adjustment && adjustment.forceRedBoxDates && adjustment.forceRedBoxDates.includes(dateStr);
                
                if (currCycle > prevCycle || isForced) {
                    allMilestones.push({ year: mc.year, month: mc.month, day: parseInt(dateStr.split('-')[2], 10), isReal: true });
                }
            }
        });`;
content = content.replace(oldMLogsForEach, newMLogsForEach);

// 3. Return allMilestones and currentCount
const oldReturn = `    return {
        redDays: actualRedDays,
        hasAnyAttendance: hasAnyAttendance,
        isSimulated: true
    };`;

const newReturn = `    return {
        redDays: actualRedDays,
        hasAnyAttendance: hasAnyAttendance,
        isSimulated: true,
        allMilestones: allMilestones,
        currentCount: { count: carryOverP, target: isDualCourse ? 17 : 9 }
    };`;
content = content.replace(oldReturn, newReturn);

// 4. Update the currentMonthLogs cycle tracking to just use the global one? No, it uses currentMC.carryFromPrev.
// Wait, the original currentMonthLogs block relies on `currentCycle`. Let's define `currentCycle` before it.
content = content.replace(/let shouldShowRedBox = false;/g, `let currentCycle = getCycle(currentMC.carryFromPrev);
                if (isNaN(currentCycle)) currentCycle = 0;
                let shouldShowRedBox = false;`);

fs.writeFileSync(path, content, 'utf8');
console.log('shared_calc.js patched');

File: patch_shared_calc.py
import re

with open('/Users/ojaeeul/Downloads/세종요리제과학원/무제 폴더/수정전/sejk 4/sejongcook-next/Sejong/SejongAttendance/public/shared_calc.js', 'r', encoding='utf-8') as f:
    content = f.read()

# I need to add `simulatedAttendances: []` to the function
# Around line 292:
#     let isSimulated = false;
#     const hasRealMilestonesThisMonth = redBoxDates.size > 0;
# We will add: `let simulatedAttendances = [];`
# Then inside `if (isValidDay && !isHoliday) {` we will push to it.

if "let simulatedAttendances = [];" not in content:
    content = content.replace("let isSimulated = false;", "let isSimulated = false;\\n    let simulatedAttendances = [];")

# Around line 365:
#             if (isValidDay && !isHoliday) {
#                 const prevSimCycle = getCycle(simTotal);
#                 simTotal += attendanceIncrement;

old_valid_day = """            if (isValidDay && !isHoliday) {
                const prevSimCycle = getCycle(simTotal);
                simTotal += attendanceIncrement;"""

new_valid_day = """            if (isValidDay && !isHoliday) {
                simulatedAttendances.push({
                    year: simDate.getFullYear(),
                    month: simDate.getMonth() + 1,
                    day: simDate.getDate()
                });
                const prevSimCycle = getCycle(simTotal);
                simTotal += attendanceIncrement;"""

if old_valid_day in content:
    content = content.replace(old_valid_day, new_valid_day)
    print("Replaced old_valid_day successfully.")
else:
    print("Could not find old_valid_day.")

# Return object replacement
old_return = """    return {
        redDays: actualRedDays,
        pureRedDays: pureRedDaysList,
        hasAnyAttendance: hasAnyAttendance,
        isSimulated: isSimulated,
        allMilestones: allMilestones,
        scheduledDate: finalScheduledDate,
        currentCount: { count: carryOverP, target: window.getCourseCycleLength(courseFilter || String(member.course), member.type) }
    };"""

new_return = """    return {
        redDays: actualRedDays,
        pureRedDays: pureRedDaysList,
        hasAnyAttendance: hasAnyAttendance,
        isSimulated: isSimulated,
        simulatedAttendances: simulatedAttendances,
        allMilestones: allMilestones,
        scheduledDate: finalScheduledDate,
        currentCount: { count: carryOverP, target: window.getCourseCycleLength(courseFilter || String(member.course), member.type) }
    };"""

if old_return in content:
    content = content.replace(old_return, new_return)
    print("Replaced old_return successfully.")
else:
    print("Could not find old_return.")

with open('/Users/ojaeeul/Downloads/세종요리제과학원/무제 폴더/수정전/sejk 4/sejongcook-next/Sejong/SejongAttendance/public/shared_calc.js', 'w', encoding='utf-8') as f:
    f.write(content)

